include/SDL3/SDL_dialog.h

```diff
@@ -0,0 +1,190 @@
+ */
+typedef struct
+{
+    const char* name;
+    const char* pattern;
+} SDL_DialogFileFilter;
+
+/**
+ * Callback used by file dialog functions.
+ *
+ * The specific usage is described in each function.
+ *
+ * If @p filelist is...
+ * - `NULL`, an error occured. Details can be obtained with SDL_GetError().
+ * - A pointer to `NULL`, the user either didn't choose any file or canceled
+ *   the dialog.
+ * - A pointer to non-`NULL`, the user chose one or more files. The argument is
+ *   a null-terminated list of pointers to C strings, each containing a path.
+ *
+ * The @p filelist argument does not need to be freed; it will automatically be
+ * freed when the callback returns.
+ *
+ * The @p filter argument is the index of the filter that was selected, or one
+ * more than the size of the list (therefore the index of the terminating NULL
+ * entry) if no filter was selected, or -1 if the platform or method doesn't
+ * support fetching the selected filter or if an error occured.
+ *
+ * \sa SDL_DialogFileFilter
+ * \sa SDL_ShowOpenFileDialog
+ * \sa SDL_ShowSaveFileDialog
+ * \sa SDL_ShowOpenFolderDialog
+ */
+typedef void(*SDL_DialogFileCallback)(void* userdata, const char* const* filelist, int filter);
+
+/**
+ * Displays a dialog that lets the user select a file on their filesystem.
+ *
+ * This function should only be invoked from the main thread.
+ *
+ * This is an asynchronous function; it will return immediately, and the result
+ * will be passed to the callback.
+ *
+ * The callback will be invoked with a null-terminated list of files the user
+ * chose. The list will be empty if the user canceled the dialog, and it will
+ * be NULL if an error occured.
+ *
+ * Note that the callback may be called from a different thread than the one
+ * the function was invoked on.
+ *
+ * If an error occurs, the error can be obtained with SDL_GetError().
+ *
+ * Depending on the platform, the user may be allowed to input paths that don't
+ * yet exist.
+ *
+ * \param callback The function to be invoked when the user selects a file and accepts, or the user cancels the dialog, or an error occurs. The first argument is a null-terminated list of C strings, representing the paths chosen by the user. The list will be empty if the user canceled the dialog, and it will be NULL if an error occured. If an error occured, it can be fetched with SDL_GetError(). The second argument is the userdata pointer passed to the function.
+ * \param userdata An optional pointer to pass extra data to the callback when it will be invoked.
+ * \param window The window to bind the dialog to. May be NULL. Not all platforms support this option.
+ * \param filters A null-terminated list of SDL_DialogFileFilter's. May be NULL. Not all platforms support this option, and platforms that do support it may allow the user to ignore the filters.
+ * \param default_location The default location to start the dialog at. May be NULL. Not all platforms support this option.
```

Comment:
You mean folder or directory here, right?

Reply:
Which line are you referring to?

Reply:
"location"

Reply:
Sure